3rdParty/fuerte/src/GeneralConnection.h

```diff
@@ -300,57 +315,79 @@ class GeneralConnection : public fuerte::Connection {
-      msg.push_back('\'');
-      shutdownConnection(Error::CouldNotConnect, msg);
+    if (_state.load() != Connection::State::Connecting) {
       return;
     }
 
     FUERTE_LOG_DEBUG << "tryConnect (" << retries << ") this=" << this << "\n";
     auto self = Connection::shared_from_this();
 
-    _proto.timer.expires_at(start + _config._connectTimeout);
-    _proto.timer.async_wait([self](asio_ns::error_code const& ec) {
-      if (!ec && self->state() == Connection::State::Connecting) {
-        // the connect handler below gets 'operation_aborted' error
-        static_cast<GeneralConnection<ST, RT>&>(*self)._proto.cancel();
-      }
-    });
-
-    _proto.connect(_config, [self, start, retries](auto const& ec) mutable {
+    _proto.connect(_config, [self, retries](asio_ns::error_code ec) mutable {
       auto& me = static_cast<GeneralConnection<ST, RT>&>(*self);
-      me._proto.timer.cancel();
+      me.cancelTimer();
       // Note that is is possible that the alarm has already gone off, in which
       // case its closure might already be queued right after ourselves!
       // However, we now quickly set the state to `Connected` in which case the
       // closure will no longer shut down the socket and ruin our success.
       if (!ec) {
+        FUERTE_LOG_DEBUG << "tryConnect (" << retries << ") established connection this=" << self.get() << "\n";
         me.finishConnect();
         return;
       }
-      FUERTE_LOG_DEBUG << "connecting failed: " << ec.message() << "\n";
+#ifdef ARANGODB_USE_GOOGLE_TESTS
+      if (me._failConnectAttempts > 0) {
+        --me._failConnectAttempts;
+      }
+#endif
+
+      FUERTE_LOG_DEBUG << "tryConnect (" << retries << "), connecting failed: " << ec.message() << "\n";
       if (retries > 0 && ec != asio_ns::error::operation_aborted) {
-        auto end = std::min(Clock::now() + me._config._connectRetryPause,
-                            start + me._config._connectTimeout);
+        FUERTE_LOG_DEBUG << "tryConnect (" << retries << "), scheduling retry operation. this=" << self.get() << "\n";
+        auto end = Clock::now() + me._config._connectRetryPause;
         me._proto.timer.expires_at(end);
         me._proto.timer.async_wait(
-            [self(std::move(self)), start, retries](auto ec) mutable {
+            [self = std::move(self), retries](asio_ns::error_code ec) mutable {
+              if (ec) {
+                FUERTE_LOG_DEBUG << "tryConnect, retry timer canceled. this=" << self.get() << "\n";
+                // timer canceled.
+                return;
+              }
               auto& me = static_cast<GeneralConnection<ST, RT>&>(*self);
-              me.tryConnect(!ec ? retries - 1 : 0, start, ec);
+              if (retries - 1 == 0) {
+                std::string msg("connecting failed: timeout");
+                FUERTE_LOG_DEBUG << "tryConnect, timeout this=" << self.get() << "\n";
+                me.shutdownConnection(Error::CouldNotConnect, msg);
```

Comment:
I do not understand why we suddenly need two places which call `shutdownConnection`, where we had only one before. But it is now too late to see this clearly. Will have another look tomorrow morning.
It seems to me that either we catch the case of exhausted retries here before calling ourselves again with `retries-1`, or we can catch it outside, but I do not think we need both.

Reply:
This comment has become obsolete with the above one.